File: app/src/main/java/vinhlong/ditagis/com/qlcln/async/QueryFeatureAsync.java
package vinhlong.ditagis.com.qlcln.async;

import android.app.ProgressDialog;
import android.content.Context;
import android.os.AsyncTask;

import com.esri.arcgisruntime.concurrent.ListenableFuture;
import com.esri.arcgisruntime.data.Feature;
import com.esri.arcgisruntime.data.FeatureQueryResult;
import com.esri.arcgisruntime.data.QueryParameters;
import com.esri.arcgisruntime.data.ServiceFeatureTable;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

import vinhlong.ditagis.com.qlcln.MainActivity;

/**
 * Created by ThanLe on 4/16/2018.
 */

public class QueryFeatureAsync extends AsyncTask<QueryParameters, List<Feature>, Void> {
    private ProgressDialog dialog;
    private Context mContext;
    private ServiceFeatureTable serviceFeatureTable;


    public QueryFeatureAsync(MainActivity mainActivity, ServiceFeatureTable serviceFeatureTable, AsyncResponse asyncResponse) {
        this.mDelegate = asyncResponse;
        mContext = mainActivity;
        this.serviceFeatureTable = serviceFeatureTable;
        dialog = new ProgressDialog(mainActivity, android.R.style.Theme_Material_Dialog_Alert);
    }

    public interface AsyncResponse {
        void processFinish(List<Feature> o);
    }

    private AsyncResponse mDelegate = null;


    @Override
    protected void onPreExecute() {
        super.onPreExecute();
        dialog.setMessage("Đang truy vấn...");
        dialog.setCancelable(false);
        dialog.show();

    }

    @Override
    protected Void doInBackground(QueryParameters... params) {
        final List<Feature> features = new ArrayList<>();
        QueryParameters queryParameters = params[0];
        final ListenableFuture<FeatureQueryResult> queryResultListenableFuture = serviceFeatureTable.queryFeaturesAsync(queryParameters, ServiceFeatureTable.QueryFeatureFields.LOAD_ALL);
        queryResultListenableFuture.addDoneListener(new Runnable() {
            @Override
            public void run() {
                try {
                    FeatureQueryResult result = queryResultListenableFuture.get();
                    Iterator iterator = result.iterator();

                    while (iterator.hasNext()) {
                        Feature feature = (Feature) iterator.next();
                        features.add(feature);
                    }
                    publishProgress(features);

                } catch (Exception e) {
                    publishProgress(features);
                }
            }
        });
        return null;
    }

    @Override
    protected void onProgressUpdate(List<Feature>... values) {
        super.onProgressUpdate(values);
        if (dialog != null && dialog.isShowing()) dialog.dismiss();
        if (values != null ) {
            mDelegate.processFinish(values[0]);

        } else {
            mDelegate.processFinish(new ArrayList<>());
        }
    }

}

File: app/src/main/java/vinhlong/ditagis/com/qlcln/async/QueryFeatureAsync.kt
package vinhlong.ditagis.com.qlcln.async

import android.app.ProgressDialog
import android.content.Context
import android.os.AsyncTask
import com.esri.arcgisruntime.data.Feature
import com.esri.arcgisruntime.data.QueryParameters
import com.esri.arcgisruntime.data.ServiceFeatureTable
import vinhlong.ditagis.com.qlcln.MainActivity
import java.util.*

/**
 * Created by ThanLe on 4/16/2018.
 */

class QueryFeatureAsync(mainActivity: MainActivity, private val serviceFeatureTable: ServiceFeatureTable, asyncResponse: AsyncResponse) : AsyncTask<QueryParameters, List<Feature>, Void>() {
    private val dialog: ProgressDialog?
    private val mContext: Context

    private var mDelegate: AsyncResponse? = null


    init {
        this.mDelegate = asyncResponse
        mContext = mainActivity
        dialog = ProgressDialog(mainActivity, android.R.style.Theme_Material_Dialog_Alert)
    }

    interface AsyncResponse {
        fun processFinish(o: List<Feature>)
    }


    override fun onPreExecute() {
        super.onPreExecute()
        dialog!!.setMessage("Đang truy vấn...")
        dialog.setCancelable(false)
        dialog.show()

    }

    override fun doInBackground(vararg params: QueryParameters): Void? {
        val features = ArrayList<Feature>()
        val queryParameters = params[0]
        val queryResultListenableFuture = serviceFeatureTable.queryFeaturesAsync(queryParameters, ServiceFeatureTable.QueryFeatureFields.LOAD_ALL)
        queryResultListenableFuture.addDoneListener {
            try {
                val result = queryResultListenableFuture.get()
                val iterator = result.iterator()

                while (iterator.hasNext()) {
                    val feature = iterator.next() as Feature
                    features.add(feature)
                }
                publishProgress(features)

            } catch (e: Exception) {
                publishProgress(features)
            }
        }
        return null
    }

    override fun onProgressUpdate(vararg values: List<Feature>) {
        super.onProgressUpdate(*values)
        if (dialog != null && dialog.isShowing) dialog.dismiss()
        if (values != null) {
            mDelegate!!.processFinish(values[0])

        } else {
            mDelegate!!.processFinish(ArrayList())
        }
    }

}

